server/src/main/java/org/opensearch/index/ShardIndexingPressure.java

```diff
@@ -0,0 +1,350 @@
+/*
+ * Copyright OpenSearch Contributors.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.index;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.admin.indices.stats.CommonStatsFlags;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.lease.Releasable;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.OpenSearchRejectedExecutionException;
+import org.opensearch.index.ShardIndexingPressureTracker.CommonOperationTracker;
+import org.opensearch.index.ShardIndexingPressureTracker.OperationTracker;
+import org.opensearch.index.ShardIndexingPressureTracker.PerformanceTracker;
+import org.opensearch.index.ShardIndexingPressureTracker.RejectionTracker;
+import org.opensearch.index.ShardIndexingPressureTracker.StatsTracker;
+import org.opensearch.index.shard.ShardId;
+import org.opensearch.index.stats.ShardIndexingPressureStats;
+import org.opensearch.index.stats.IndexingPressurePerShardStats;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Shard Indexing Pressure is a framework level artefact build on top of IndexingPressure to track incoming indexing request, per shard.
+ * The interfaces provided by this class will be used by Transport Action layers to start accounting for an incoming request.
+ * Interfaces returns Releasable which when triggered will release the acquired accounting tokens values and also
+ * perform necessary actions such as throughput evaluation once the request completes.
+ * Consumers of these interfaces are expected to trigger close on releasable, reliably for consistency.
+ *
+ * Overall ShardIndexingPressure provides:
+ * 1. Memory Accounting at shard level. This can be enabled/disabled based on dynamic setting.
+ * 2. Memory Accounting at Node level. Tracking is done using the IndexingPressure artefacts to support feature seamless toggling.
+ * 3. Interfaces to access the statistics for shard trackers.
+ */
+public class ShardIndexingPressure extends IndexingPressure {
```

Comment:
I don't think the core logic here got updated more than once in this PR. It exposes 4 public methods today, which are on the lines of ``IndexingPressure``,  and takes an additional ``shard`` field parameter. Each of these methods are for different shard operations i.e. coordinating, primary-local, primary and replica. So it is having the similar responsibilities as of ``IndexingPressure`` today.

Also from the inheritance perspective it is not fake, but in actual it is extending the properties/attributes of ``IndexingPressure`` so that values are updated appropriately, irrespective of the feature flag being on/off. Not really sure why you think this is composition since ``IndexingPressure`` is not part of ``ShardIndexingPressure``.

I have thought this few times and have taken a conscious decision of not re-using the methods from ``IndexingPressure`` as the logic in total is not A+B type. It adds more complexity and tight coupling between the two classes when updating values during exceptions and release scenarios.

Moreover, we also envision with rollout of ``ShardIndexingPressure ``, we will eventually be able to deprecate ``IndexingPressure``.

Reply:
Thats exactly my concern, the moment you deprecate `IndexingPressure ` the `ShardIndexingPressure ` is gonna become a monolith. Inheritance is just not about using attributes but also extending methods, currently its just doing one part and duplicating the parent logic. Handling node level and shard level at one place makes the class overloaded which could have fanned out

Reply:
> I have thought this few times and have taken a conscious decision of not re-using the methods from IndexingPressure as the logic in total is not A+B type. It adds more complexity and tight coupling between the two classes when updating values during exceptions and release scenarios.

I would be happy to understand those cases and see if I could help with loose coupling. Let me know

Reply:
``ShardIndexingPressure`` is not handling the core logic of node or shard level evaluation, but it is being delegating to the lower level constructs ``ShardIndexingPressureMemoryManager`` and ``ShardIndexingPressureStore`` with right responsibilities and hence avoiding it to become monolith. Also from the responsibilities side, it is exposing similar interfaces to clients as mentioned above, hence following the single responsibility principles. 

I will chose to keep the logic separate, as there is not much reusability but tight coupling, given the logic is not additive in nature.

Reply:
I would humbly disagree. The reason I recommended you to extend was based on my understanding that the logic would be reused/fanned and delegated correctly at different levels. If you don't think thats possible I don't think extending `IndexingPressure` is the right design given there are hardly any commonalities with not even a single logic extended or overridden

Reply:
Extension was a choice and we wanted to re-use for attributes as ``IndexingPressure`` attributes. It is required so that the ``ShardIndexingPressure`` as a feature can be toggled, which the state of ``IndexingPressure`` continues. Not using methods is more an implementation choice to loose the coupling between these two classes. It also gives us frame to build ``ShardIndexingPressure`` and its underlying constructs in isolation, with liberty to eventually deprecate ``IndexingPressure``.

Can you please be more specific with the concern here?

Reply:
1. Code duplication in sub-class when an extension is possible
2. Too much logic in `ShardIndexingPressure` which is a maintenance overhead. 
As a reviewer I got lost so will others looking at the code. If you don't want to improve on this suggestion it's still fine

Reply:
> Code duplication in sub-class when an extension is possible

There is no code in ``ShardIndexingPressure`` which could be offloaded to ``IndexingPressure`` directly. Core logic for node limit check is already segregated and offloaded to ``ShardIndexingPressureMemoryManager``. For the node level counter updates in ``ShardIndexingPressure``, since logic is not additive, still prefer to not use methods, as its not directly consumable. Changes in ``IndexingPressure`` will unnecessarily increase the coupling which I want to avoid. 

> Too much logic in ShardIndexingPressure which is a maintenance overhead.

I will be happy to take more inputs and rather focus on segregating the ``ShardIndexingPressure``  and making it simpler. I dont think we should be blocked on reusing ``IndexingPressure`` just for counter updates, as its unnecessary overhead.